### PR TITLE
feat(store)!: replace structuredClone with custom deep clone

### PR DESCRIPTION
**Pull request by Marvin-Brouwer**
Replaced `structuredClone` with a custom `deepClone` implementation that preserves functions and symbol-keyed properties, which are important for branded types and methods on state objects. The store now passes the live state reference to the setter (not a clone), and caches frozen snapshots between updates for better memoization support.

## Definition Of Done

- [x] Meets acceptance criteria of story / Fixed bug
- [x] Added new tests
- [x] All tests pass (`pnpm test`)
- [x] Checked regression
  - Build passes
  - Tests still pass
  - Linting still passes
- [x] Updated documentation
  - TSDoc added/updated for any changed public APIs

## Summary of Changes

**New utilities** (`deepClone.mts`):
- `deepClone()` - Deep clones objects, arrays, Date, Map, Set while preserving functions by reference and symbol-keyed properties. Handles cycles via a WeakMap.
- `deepFreeze()` - Recursively freezes objects and arrays in place, skipping class instances and Map/Set which can't be meaningfully frozen.

**Store behavior changes**:
- The setter now receives the live state reference instead of a clone, allowing direct mutations like `s.pad.aces = score`.
- Snapshots are cached between updates and invalidated on each `update()` call, so `store.value === store.value` between updates (useful for downstream memoization).
- Symbol-keyed properties and functions in state are now preserved through updates.

**Hash function improvements**:
- Functions are now explicitly dropped during hashing (they're identity-only and can't be meaningfully hashed).
- Symbol-keyed properties are included in the hash calculation.

All existing tests pass, and comprehensive new tests cover edge cases like cycles, symbol properties, functions in state, and the caching behavior.

https://claude.ai/code/session_01ADz2mWihKMecyw27qr1Hmm

### CODE REVIEW

File: packages/store/src/_module/store.mts [hunk 1]
```diff
@@ -9,3 +9,4 @@
  */
 
 export { createStore, type Store, type StoreEvent } from '../store.mts'
+export { deepClone, deepFreeze } from '../deepClone.mts'
```
Comment (Marvin-Brouwer) [inline packages/store/src/_module/store.mts]: I don't think `deepFreeze` has any benefit outside of the internals

File: packages/store/src/deepClone.mts [hunk 2]
```diff
@@ -0,0 +1,123 @@
+/**
+ * Deep-clones a value without going through `structuredClone`.
+ *
+ * Why not `structuredClone`? Two reasons:
```
Comment (Marvin-Brouwer) [inline packages/store/src/deepClone.mts]: I don't think a user of this library cares about why we moved away from structured clone.
Please keep the TS doc useful to the consumer.

File: packages/store/src/hash.mts [hunk 3]
```diff
@@ -1,20 +1,32 @@
 function hashReplacer(_key: string, value: unknown): unknown {
+	// eslint-disable-next-line unicorn/no-null
 	if (value === null) return value
 	if (value === undefined) return value
+	// Functions are identity-only and can't be hashed. Drop them.
+	if (typeof value === 'function') return undefined
```
Comment (Marvin-Brouwer) [inline packages/store/src/hash.mts]: > Functions are identity-only and can't be hashed. Drop them.
Then why don't return the identity here?

File: packages/store/src/store.mts [hunk 4]
```diff
@@ -56,58 +67,32 @@ class StoreImpl<TState extends StateType | Array<StateType>> extends EventTarget
...
-		else {
-			this.#state = result as TState
-		}
-
-		const frozenState = this.#isObject
-			? Object.freeze(this.#clone(this.#state))
-			: this.#state as Readonly<TState>
+		this.#snapshot = undefined
 
+		const frozenState = this.value
```
Comment (Marvin-Brouwer) [inline packages/store/src/store.mts]: This should be deepFrozen

File: packages/store/src/deepClone.mts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+ * What gets cloned: plain objects, arrays, `Date`, `Map`, `Set`, and any symbol-keyed properties on them. Cycles are handled.
+ *
+ * What stays shared by reference: functions and class instances (anything whose prototype isn't `Object.prototype` or `null`). There's no general way to reconstruct those, so the original reference is reused.
+ *
+ * Useful when you want to detach a frozen `store.value` into a mutable copy:
+ *
+ * ```ts
+ * import { deepClone } from '@rooted/store'
+ *
+ * const mine = deepClone(store.value)
```
Comment (Marvin-Brouwer) [inline packages/store/src/deepClone.mts]: I'm not sure this is a valuable example.
The store.value is completely immutable so wouldn't deepclone also be immutables?

Also, this is just a deepClone, it's exported, I don't think it necessarily HAS to be used with a store.

File: packages/store/src/deepClone.mts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+	if (value instanceof Date) {
+		Object.freeze(value)
+		return value
+	}
+
+	if (value instanceof Map || value instanceof Set) {
+		// Freezing a Map/Set doesn't stop mutation through its own methods,
+		// so don't pretend. Leave them mutable.
+		return value
```
Comment (Marvin-Brouwer) [inline packages/store/src/deepClone.mts:103]: Incorrect, we Recursively make this immutable in the Type, so we should also do it here.
